gazelle_haskell_modules/lang.go

```diff
@@ -171,10 +173,57 @@ func (*gazelleHaskellModulesLang) Embeds(r *rule.Rule, from label.Label) []label
 
 func (*gazelleHaskellModulesLang) Resolve(c *config.Config, ix *resolve.RuleIndex, rc *repo.RemoteCache, r *rule.Rule, imports interface{}, from label.Label) {
 	hmc := c.Exts[gazelleHaskellModulesName].(Config)
-	if isNonHaskellModule(r.Kind()) {
-		setNonHaskellModuleDeps(&hmc, c.RepoRoot, ix, r, imports.(*HRuleImportData), from)
-	} else {
-		setHaskellModuleDeps(ix, r, imports.(*HModuleImportData), from)
+	switch expr := imports.(type) {
+	// This happens only when one run gazelle_cabal and gazelle_haskell_modules in one pass as the the gazelle_binary.
+	case gazelle_cabal.ImportData:
+		// We fully generate the rule that gazele_cabal would have produced.
+		gazelle_cabal.RunResolve(c, ix, rc, r, expr, from)
```

Comment:
I need some help understanding the flow. Is this case entered anytime a `haskell_binary/library/test` rule is updated or created by `gazelle_cabal`?

If so, how are `haskell_module` rules generated from this rule?

Reply:
This case is entered whenever the `Resolve` function is called on rules that come from the same gazelle run (so rules where the `imports` where constructed by `gazelle_cabal` and not by the haskell lib `himportscan` included in `gazelle_haskell_modules`).
Regarding the construction of `haskell_modules` rules, it is performed just like for other rules by `setNonHaskellModuleDeps`.

Reply:
This code is calling `RunResolve`, which apparently is a duplicate of the `Resolve` callback of `gazelle_cabal`. Isn't that callback invoked by `gazelle` implicitly?